### PR TITLE
SD-1327 Refactor quasar web to expose routes as a program in Free

### PR DESCRIPTION
**Pull request by drostron**

### CODE REVIEW

File: core/src/main/scala/quasar/errors.scala [hunk 1]
```diff
@@ -51,6 +52,12 @@ object Errors {
     liftMT[Task, G]
   }
 
+  /** Flatten by inserting the `E` into the failure case of `Task` */
+  def flatten[E:Show] = new (ETask[E,?] ~> Task) {
+    def apply[A](t: ETask[E,A]): Task[A] =
+      t.fold(e => Task.fail(new RuntimeException(e.shows)), Task.now).join
+  }
+
```
Comment (sellout) [inline core/src/main/scala/quasar/errors.scala]: This scares me. Why do we need this?

File: core/src/test/scala/quasar/fs/InMemoryArbitrary.scala [hunk 2]
```diff
@@ -30,3 +30,5 @@ trait InMemoryArbitrary {
   implicit val arbitraryInMemState: Arbitrary[InMemState] =
     Arbitrary(arb[Map[AFile, Vector[Data]]].map(InMemState(0, _, Map(), Map(), Map(), Map())))
 }
+
+object InMemoryArbitrary extends InMemoryArbitrary
```
Comment (sellout) [inline core/src/test/scala/quasar/fs/InMemoryArbitrary.scala]: How is it even possible to make a file without a final newline?

File: web/src/main/scala/quasar/api/Http4sUtils.scala [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+import shapeless._
+import shapeless.Nat
+import shapeless.ops.nat._
+import shapeless.nat._
+
+object Http4sUtils {
+
+//  sealed trait BindingType
+//  case object HTTP extends BindingType
+//  case object Local extends BindingType
```
Comment (sellout) [inline web/src/main/scala/quasar/api/Http4sUtils.scala]: Are these obsolete or aspirational?

File: core/src/test/scala/quasar/fs/InMemoryArbitrary.scala (hunk 2 above)
Comment (drostron) [inline core/src/test/scala/quasar/fs/InMemoryArbitrary.scala]: Haha, exactly!

File: core/src/main/scala/quasar/errors.scala (hunk 1 above)
Comment (jedesah) [inline core/src/main/scala/quasar/errors.scala]: Basically, we need this at the edge of the world. We need it for Http Streaming and maybe somewhere else. I think it's probably a good idea to move it to `web` as I don't see a good reason why it should be defined (or used!) in `core`.

File: web/src/main/scala/quasar/api/Http4sUtils.scala (hunk 3 above)
Comment (jedesah) [inline web/src/main/scala/quasar/api/Http4sUtils.scala]: We will probably need them if we ever want to support restarting the server on a new port again in quasar-advanced.

Comment (jedesah) [inline web/src/main/scala/quasar/api/Http4sUtils.scala]: But I think we can get rid of it for now.

**Comment by drostron** [comment]
@sellout, pushed a few commits addressing comments.

**Comment by sellout** [comment]
Ok, this looks good. The new approach is nice.

The review was a bit more cursory than I like, but with three of you already collaborating, I’m not too concerned about that.

**Pull request merged**